Add implementation of fingerprint orientation

andrew-farries opened the pull request:
This PR contains the implementation of the orientation step, and some associated new data structures and tests.

SourceAFIS uses a random(ish) sampling of neighbours but we're using a hard coded set, the same for each pixel. We're not sure how much impact that has on the result.

Currently, the smoothing step of orientation is not performed. The results don't look fantastic but it's also a bit hard to tell with the current visualisation.

Commit:
Add new array types

* 2D arrays of booleans
* 2D arrays of 16 bit integers
* 2D arrays of 'PointF's

Commit:
Add scalar multiplication function

Commit:
Add vector addition for PointFs

Commit:
Orientation implementation

Commit:
Add tests to visualise pixel mask and orientation

Commit:
Add -lm to linker flags

Link the math library.

fatboab commented on Sources/Extraction/Filters/HillOrientation.c:
```diff
@@ -1,26 +1,183 @@
+    printf("\n");
+  }
+}
+*/
```
Do we need commented out code?

fatboab commented on Sources/Extraction/Filters/HillOrientation.c:
```diff
@@ -1,26 +1,183 @@
+#include "General/Array.h"
 
-HillOrientation HillOrientation_Construct(void)
+#define MIN(X, Y) (((X) < (Y)) ? (X) : (Y))
```
We need to have a common include for these, this is the second time I have read this code today!

fatboab commented on Sources/Tests/Extraction/Filters/Test_HillOrientation.c:
```diff
@@ -0,0 +1,135 @@
+      }
+
+    }
+  }
```
Please free memory

Commit:
Move LFLAGS later in to command line

Linker flags should come near the end of the command line so that linking
works. See:

http://stackoverflow.com/questions/5817269/does-the-order-of-l-and-l-options-in-the-gnu-linker-matter

Commit:
Merge branch 'master' into orientation

fatboab commented on Sources/Extraction/Filters/HillOrientation.h:
```diff
@@ -8,27 +8,10 @@
-typedef struct NeighborInfo NeighborInfo;
-typedef struct HillOrientation HillOrientation;
+UInt16Array2D HillOrientation_Detect(FloatArray2D image, Size imageDimensions, BinaryMap * blockMask, BlockMap * blocks);
+BoolArray2D HillOrientation_BlockMapToPixelMask(Size imageDimensions, BinaryMap * blockMask, BlockMap * blocks);
```
Do we really need to expose this for everyone? I pressume that you are using it for testing?

fatboab commented on Sources/Extraction/Filters/HillOrientation.c:
```diff
@@ -1,26 +1,183 @@
 
-void /*UInt8Array2D*/ HillOrientation_Detect(const HillOrientation *me, const FloatArray2D *image, const BinaryMap *mask, const BlockMap *blocks)
+BoolArray2D HillOrientation_BlockMapToPixelMask(Size imageDimensions, BinaryMap * blockMask, BlockMap * blocks) {
+    const int blockHeight = 15; //blocks->pixelCount.height / blocks->blockCount.height;
```
Why is this commented out? Do you really need to duplicate the blockSize?

fatboab commented on Sources/Extraction/Filters/HillOrientation.c:
```diff
@@ -1,26 +1,183 @@
 }
 
-void /*UInt8Array2D*/ HillOrientation_Detect(const HillOrientation *me, const FloatArray2D *image, const BinaryMap *mask, const BlockMap *blocks)
+BoolArray2D HillOrientation_BlockMapToPixelMask(Size imageDimensions, BinaryMap * blockMask, BlockMap * blocks) {
```
Cann't this be static? (infact this could be the same for most of these methods?)

fatboab commented on Sources/Extraction/Filters/HillOrientation.c:
```diff
@@ -1,26 +1,183 @@
+            int bottomLeftX = blocks->blockAreas.corners.allX.data[blockX];
+            int bottomLeftY = blocks->blockAreas.corners.allY.data[blockY];
+            for (int x = bottomLeftX; x < bottomLeftX + blockWidth; ++x) {
+                for ( int y = bottomLeftY; y < bottomLeftY + blockHeight; ++y ) {
```
Mixing of bracket styles

fatboab commented on Sources/Extraction/Filters/HillOrientation.c:
```diff
@@ -1,26 +1,183 @@
+            int bottomLeftY = blocks->blockAreas.corners.allY.data[blockY];
+            for (int x = bottomLeftX; x < bottomLeftX + blockWidth; ++x) {
+                for ( int y = bottomLeftY; y < bottomLeftY + blockHeight; ++y ) {
+                    pixelMask.data[x][y] = blockIsInImage;
```
How are you not over-writing past the end of the array if your image is not an exact multiple of 15?
I don't think this if (x < imageDiemnsions.width && (y < imageDiemnsions.height)) is acceptable...

fatboab commented on Sources/Extraction/Filters/HillOrientation.c:
```diff
@@ -1,26 +1,183 @@
+    //TODO smooth
+
+    PointFArray2D_Destruct(&pixelDirections);
+    //TODO free pixelMask
```
Yup I agree you are leaking memory here

fatboab replied:
What about BlockDirection too?

fatboab commented on Sources/Extraction/Filters/HillOrientation.c:
```diff
@@ -1,26 +1,183 @@
+    { 3, 0 }, { 3, 1  }, { 3, 2 }, { 3, 3 },
+    { -5, 0 }, { -5, 5 }, { 0, 5 }, { 5, 5 }
+};
+
```
Perosnally I'd have done const Point neighbors[]
folloowed by 
#define NUM_VECTORS  sizeof(neighbors)/sizeof(neighbors[0]) 
and then you can add to these as you like

fatboab commented on Sources/Extraction/Filters/HillOrientation.c:
```diff
@@ -1,26 +1,183 @@
+  }
+} */
+
+PointFArray2D HillOrientation_AccumulateDirections(FloatArray2D input, int imageHeight, int imageWidth, BoolArray2D pixelMask)
```
Why not pass in Size to start off with and then you don;t need to re-create the Size object thingy

fatboab commented on Sources/Extraction/Filters/HillOrientation.c:
```diff
@@ -1,26 +1,183 @@
+                float pixelValue = input.data[x][y];
+                float neighborValue = input.data[neighbor.x][neighbor.y];
+
+                float strength = abs(pixelValue - MAX(neighborValue, antiNeighborValue));
```
Do you mean fabs? Also why abs it? I can't see this in the original code? Also the original code only allowed strengths > 0, is yours an improvement then?

fatboab commented on Sources/Extraction/Filters/HillOrientation.c:
```diff
@@ -1,26 +1,183 @@
+                Point antiPoint = (Point) { .x = -neighbors[i].x, .y = -neighbors[i].y };
+                Point antiNeighbor = Calc_Add2Points(&antiPoint, &(Point) { .x = x, .y = y });
+
+                if(!IsInsideImageAndMask(imageHeight, imageWidth, pixelMask, neighbor.x, neighbor.y) || !IsInsideImageAndMask(imageHeight, imageWidth, pixelMask, antiNeighbor.x, antiNeighbor.y))
```
We observed that this is new too?

fatboab commented on Sources/Extraction/Filters/HillOrientation.c:
```diff
@@ -1,26 +1,183 @@
+
+                float strength = abs(pixelValue - MAX(neighborValue, antiNeighborValue));
+
+                PointF contribution = Calc_Scalar_Multiply(strength, neighbor);
```
Can see that this is the same as the original algorithm, however don;t understand, if the anti-neighbour was used, then surely we should adding that one? as the direction will be opposite?

olane replied:
Just had a long conversation with Arun about this.

Summary: you're right, this is wrong, we should be calculating the orientation and adding that to the contribution.

fatboab commented on Sources/Extraction/Filters/HillOrientation.c:
```diff
@@ -1,26 +1,183 @@
-        .smoothingRadius = 1     /* Upper = 3 */
-    };
+    const PointFArray2D blockOrientations = PointFArray2D_Construct(blocks->blockCount.width, blocks->blockCount.height);
+    const int blockHeight = blocks->pixelCount.height / blocks->blockCount.height;
```
What about rounding? Is this correct truncation?

fatboab commented on Sources/Extraction/Filters/HillOrientation.c:
```diff
@@ -1,26 +1,183 @@
+            int bottomLeftY = blocks->blockAreas.corners.allY.data[blockY];
+            for (int x = bottomLeftX; x < bottomLeftX + blockWidth; ++x) {
+                for ( int y = bottomLeftY; y < bottomLeftY + blockHeight; ++y ) {
+                    blockOrientations.data[blockX][blockY] = Calc_Add2PointsF(&blockOrientations.data[blockX][blockY], &pixelOrientations->data[x][y]);
```
Are we really keeping inside the array pixelOrientations->data[x][y] ?

fatboab commented on Sources/Extraction/Filters/HillOrientation.c:
```diff
@@ -1,26 +1,183 @@
+    PointFArray2D pixelDirections = HillOrientation_AccumulateDirections(image, imageDimensions.height, imageDimensions.width, pixelMask);
+    PointFArray2D blockDirections = HillOrientation_SumBlocks(&pixelDirections, blockMask, blocks);
+    UInt16Array2D angles = HillOrientation_DirectionsToAngles(blockDirections, blockMask);
+    //TODO smooth
```
We think this comment should go before working out DirectionsToAngles?

OK Change of heart
Call your function HillOrientation_Smooth
But make it return it's input then leave your todo there?

fatboab commented on Sources/Extraction/Filters/HillOrientation.c:
```diff
@@ -1,26 +1,183 @@
+    return (PointFArray2D) { NULL };
+}
+
+UInt16Array2D HillOrientation_DirectionsToAngles(PointFArray2D directions, BinaryMap *mask)
```
Why not make HillOrientation_DirectionsToAngles pass in a pointer to directions?

fatboab commented on Sources/Extraction/Filters/HillOrientation.c:
```diff
@@ -1,26 +1,183 @@
+
+            double angle = atan2(directions.data[x][y].y, directions.data[x][y].x);
+            //multiply by 2 to collapse opposite angles onto each other
+            angles.data[x][y] = (uint16_t)(angle / 3.1415f * 360 * 2 + 90) % 360;
```
I tried an angle or 45 & 225 
and 90/270 
and I could not make it give me the expected answer
I think that Radians-> Degrees is (radians/PI) \* 180
So the calc I was checking was (degrees \* 2 \* 2) % 360 ie ignoring the add 90degrees
Cheers David
